### PR TITLE
Fix #15; logging top ten lowest loss samples

### LINKED ISSUE #15: Instrument DeepSpeech to display epoch's top N most confident results with WER!=0 orderer by WER

### PR DESCRIPTION
**Pull request by tilmankamp**
@kdavis-mozilla : First draft of the top ten lowest loss report. Please take a brief look and tell, if this is what you intended.

### CODE REVIEW

**Comment by kdavis-mozilla** [comment]
@tilmankamp Ok I did a quick look over, not really a review per-say.

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1239,16 +1271,21 @@
...
     "        # Loop over the batches\n",
-    "        for batch in range(int(ceil(float(total_batches)/len(available_devices)))):\n",
+    "        for batch in range(int(ceil(float(total_batches) / len(available_devices)))):\n",
     "            # Compute the average loss for the last batch\n",
-    "            _, batch_avg_loss = session.run([apply_gradient_op, tower_loss], feed_dict_train)\n",
+    "            session.run(apply_gradient_op, feed_dict_train)\n",
     "\n",
     "            # Add batch to total_accuracy\n",
-    "            total_accuracy += session.run(accuracy, feed_dict_train)\n",
+    "            total_accuracy += session.run(tower_accuracies[-1], feed_dict_train)\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: This mimics the old behavior where just the last tower's accuracy was used. I agree with you on using the average of all tower accuracies instead.

File: DeepSpeech.ipynb [hunk 2]
```diff
@@ -1386,7 +1428,7 @@
     "            'total_batches_validation': data_sets.dev.total_batches, \\\n",
     "            'total_batches_test': data_sets.test.total_batches, \\\n",
     "            'data_set': { \\\n",
-    "                'name': ds_importer\n",
+    "                'name': ds_importer \\\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Aesthetics only: For the (added) backslash not being visually glued to ```ds_importer``` like ```ds_importer\```. I prefer ```ds_importer \```.

**Comment by tilmankamp** [comment]
@kdavis-mozilla : I think it's time for having a full review on this.

File: DeepSpeech.ipynb [hunk 3]
```diff
@@ -1258,8 +1308,13 @@
     "        \n",
     "        # Print progress message\n",
     "        if epoch % display_step == 0:\n",
-    "            print \"Epoch:\", '%04d' % (epoch+1), \"avg_cer=\", \"{:.9f}\".format((total_accuracy / total_batches))\n",
-    "            _, last_train_wer = print_wer_report(session, \"Training\", tower_decodings, tower_labels)\n",
+    "            print \"Epoch:\", '%04d' % (epoch+1), \"avg_cer=\", \"{:.9f}\".format((total_accuracy / batches_per_device))\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Right.

File: DeepSpeech.ipynb [hunk 4]
```diff
@@ -1232,23 +1276,29 @@
...
     "    last_validation_wer = 0.0\n",
     "    \n",
+    "    \n",
     "    # Loop over the data set for training_epochs epochs\n",
     "    for epoch in range(training_iters):\n",
     "        # Define total accuracy for the epoch\n",
-    "        total_accuracy = 0\n",
+    "        total_accuracy = 0.0\n",
     "        \n",
     "        # Validation step\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: > thus the WER will simply reflect the WER of the random initialization of the system
Exactly this was intended. To be able to compare. However: I'll move it down.

File: DeepSpeech.ipynb [hunk 5]
```diff
@@ -1163,13 +1183,28 @@
...
+    "        items.sort(key=lambda a: a[2])\n",
+    "        for a in items:\n",
+    "            print \"-\" * 80\n",
+    "            print \" - WER:    %f\" % a[2]\n",
+    "            print \" - source: \\\"%s\\\"\" % a[0]\n",
+    "            print \" - result: \\\"%s\\\"\" % a[1] \n",
+    "            print \" - loss:   %f\" % a[3]\n",
+    "    print \"#\" * 80\n",
+    "    print\n",
     "    return items, mean"
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Improved this now by splitting the method up into a calculation and printing part - which is also much cleaner from a design point of view.

**Comment by tilmankamp** [comment]
@kdavis-mozilla : Time for the next round.

**Comment by kdavis-mozilla** [comment]
@tilmankamp You want to merge or should I
